File: app/src/main/java/com/example/islandlyrics/MediaMonitorService.java
package com.example.islandlyrics;

import android.content.ComponentName;
import android.content.Context;
import android.content.SharedPreferences;
import android.media.MediaMetadata;
import android.media.session.MediaController;
import android.media.session.MediaSessionManager;
import android.media.session.PlaybackState;
import android.service.notification.NotificationListenerService;
import android.util.Log;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class MediaMonitorService extends NotificationListenerService {
    private static final String TAG = "MediaMonitorService";
    private static final String PREFS_NAME = "IslandLyricsPrefs";
    private static final String PREF_WHITELIST = "whitelist_json";
    
    private MediaSessionManager mMediaSessionManager;
    private ComponentName mComponentName;
    private SharedPreferences mPrefs;
    
    private final Set<String> mAllowedPackages = new HashSet<>();

    @Override
    public void onCreate() {
        super.onCreate();
        mPrefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        // Load initial whitelist
        loadWhitelist();
        // Register pref listener
        mPrefs.registerOnSharedPreferenceChangeListener(mPrefListener);
    }

    @Override
    public void onListenerConnected() {
        super.onListenerConnected();
        Log.d(TAG, "onListenerConnected");
        mMediaSessionManager = (MediaSessionManager) getSystemService(Context.MEDIA_SESSION_SERVICE);
        mComponentName = new ComponentName(this, MediaMonitorService.class);

        mMediaSessionManager.addOnActiveSessionsChangedListener(mSessionsChangedListener, mComponentName);
        
        // Initial check
        List<MediaController> controllers = mMediaSessionManager.getActiveSessions(mComponentName);
        updateControllers(controllers);
    }

    @Override
    public void onListenerDisconnected() {
        super.onListenerDisconnected();
        if (mMediaSessionManager != null) {
            mMediaSessionManager.removeOnActiveSessionsChangedListener(mSessionsChangedListener);
        }
    }
    
    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mPrefs != null) {
            mPrefs.unregisterOnSharedPreferenceChangeListener(mPrefListener);
        }
    }
    
    private final SharedPreferences.OnSharedPreferenceChangeListener mPrefListener = 
        (sharedPreferences, key) -> {
            if (PREF_WHITELIST.equals(key)) {
                loadWhitelist();
                recheckSessions();
            } else if ("service_enabled".equals(key)) {
                recheckSessions();
            }
        };

    private void recheckSessions() {
        if (mMediaSessionManager != null) {
            try {
                updateControllers(mMediaSessionManager.getActiveSessions(mComponentName));
            } catch (SecurityException e) {
                AppLogger.getInstance().log(TAG, "Error refreshing sessions: " + e.getMessage());
            }
        }
    }

    private void loadWhitelist() {
        // Use Helper to get only Enabled packages
        Set<String> set = WhitelistHelper.getEnabledPackages(this);
        mAllowedPackages.clear();
        mAllowedPackages.addAll(set);
        AppLogger.getInstance().log(TAG, "Whitelist updated: " + mAllowedPackages.size() + " enabled apps.");
    }

    private final java.util.List<android.media.session.MediaController> mActiveControllers = new java.util.ArrayList<>();

    private final android.media.session.MediaSessionManager.OnActiveSessionsChangedListener mSessionsChangedListener = 
            this::updateControllers;

    private final android.os.Handler mHandler = new android.os.Handler(android.os.Looper.getMainLooper());
    private final Runnable mStopRunnable = () -> {
        AppLogger.getInstance().log(TAG, "Stopping service after debounce.");
        android.content.Intent intent = new android.content.Intent(MediaMonitorService.this, LyricService.class);
        intent.setAction("ACTION_STOP");
        startService(intent);
    };

    private void updateControllers(List<MediaController> controllers) {
        // AppLogger.getInstance().log(TAG, "Segments changed.");
        
        // CHECK MASTER SWITCH
        boolean isServiceEnabled = mPrefs.getBoolean("service_enabled", true);
        if (!isServiceEnabled) {
             AppLogger.getInstance().log(TAG, "Master Switch OFF. Ignoring updates.");
             mActiveControllers.clear();
             LyricRepository.getInstance().updatePlaybackStatus(false);
             return;
        }
        
        mActiveControllers.clear();
        if (controllers != null) {
            for (MediaController controller : controllers) {
                if (mAllowedPackages.contains(controller.getPackageName())) {
                    mActiveControllers.add(controller);
                    // Re-register callbacks to ensure we catch state changes
                    // Note: In production, careful not to double-register. 
                    // MediaSessionManager usually returns new objects.
                    controller.registerCallback(new MediaController.Callback() {
                        @Override
                        public void onPlaybackStateChanged(PlaybackState state) {
                            checkServiceState();
                        }
                        @Override
                        public void onMetadataChanged(MediaMetadata metadata) {
                            updateMetadataIfPrimary(controller);
                        }
                    });
                }
            }
        }
        
        // Initial check
        checkServiceState();
        
        // Force update metadata from the CURRENT primary
        MediaController primary = getPrimaryController();
        if (primary != null) {
            updateMetadataIfPrimary(primary);
        }
    }
    
    private void checkServiceState() {
        MediaController primary = getPrimaryController();
        boolean isPlaying = primary != null && 
                            primary.getPlaybackState() != null &&
                            primary.getPlaybackState().getState() == PlaybackState.STATE_PLAYING;

        if (isPlaying) {
            // Cancel any pending stop
            mHandler.removeCallbacks(mStopRunnable);
            
            // Start/Update Service
            // AppLogger.getInstance().log(TAG, "Starting/Keeping service for: " + primary.getPackageName());
            android.content.Intent intent = new android.content.Intent(MediaMonitorService.this, LyricService.class);
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
                startForegroundService(intent);
            } else {
                startService(intent);
            }
            
            // Sync State
            LyricRepository.getInstance().updatePlaybackStatus(true);
            
        } else {
            // Debounce Stop
            // Only post if not already posting? 
            // Better: remove previous and post new to reset timer? 
            // No, just ensure we don't have duplicates.
            // But if we are already stopping, just let it be.
            // If we transitioned Play -> Pause, we trigger this.
            
            // If already posted, do nothing? Or defer?
            // Simple approach: Remove and Post (Reset timer)
            mHandler.removeCallbacks(mStopRunnable);
            mHandler.postDelayed(mStopRunnable, 500);
            
            LyricRepository.getInstance().updatePlaybackStatus(false);
        }
    }

    private MediaController getPrimaryController() {
        // Priority 1: Playing
        for (MediaController c : mActiveControllers) {
            PlaybackState state = c.getPlaybackState();
            if (state != null && state.getState() == PlaybackState.STATE_PLAYING) {
                return c;
            }
        }
        // Priority 2: First in list (or most recent?) - API list order usually implies recency
        if (!mActiveControllers.isEmpty()) {
            return mActiveControllers.get(0);
        }
        return null; // None active
    }

    private void updateMetadataIfPrimary(MediaController controller) {
        MediaController primary = getPrimaryController();
        if (primary == null) return;
        
        // Only update if this controller IS the primary one
        if (controller.getPackageName().equals(primary.getPackageName())) {
             MediaMetadata metadata = controller.getMetadata();
             if (metadata != null) {
                 extractMetadata(metadata, controller.getPackageName());
             }
        }
    }

    private void extractMetadata(MediaMetadata metadata, String pkg) {
        if (metadata == null) return;
        
        String rawTitle = metadata.getString(MediaMetadata.METADATA_KEY_TITLE);
        String rawArtist = metadata.getString(MediaMetadata.METADATA_KEY_ARTIST);
        
        AppLogger.getInstance().log("Meta-Debug", "📦 RAW Data from [" + pkg + "]:");
        AppLogger.getInstance().log("Meta-Debug", "   ↳ Title : " + (rawTitle == null ? "null" : rawTitle));
        AppLogger.getInstance().log("Meta-Debug", "   ↳ Artist: " + (rawArtist == null ? "null" : rawArtist));

        String finalTitle = rawTitle;
        String finalArtist = rawArtist;
        String finalLyric = null; // Default to null

        // Target specific apps known to use the "Car Bluetooth" hack
        boolean isTencentBase = pkg.contains("tencent") || pkg.contains("miui.player");

        if (isTencentBase) {
            // Check if it looks like the Car Protocol (Artist field acts as container)
            // Check if it looks like the Car Protocol (Artist field acts as container)
            // or if we just want to try parsing it.
            if (rawArtist != null) {
                String separator = null;
                int splitIndex = -1;
                int offset = 0;

                // STRATEGY 1: Look for Standard " - " (Space Hyphen Space)
                // Use Last Index here (Safe for "Anti-Hero - Taylor Swift")
                if (rawArtist.contains(" - ")) {
                    separator = " - ";
                    splitIndex = rawArtist.lastIndexOf(separator);
                    offset = separator.length();
                } 
                // STRATEGY 2: Fallback to Tight "-" (Hyphen)
                // Use FIRST Index here to protect Artist names like "HOYO-MiX"
                // Assumption: "Song-HOYO-MiX" -> Split at 1st hyphen -> Title="Song", Artist="HOYO-MiX"
                else if (rawArtist.contains("-")) {
                    separator = "-";
                    splitIndex = rawArtist.indexOf(separator);
                    offset = separator.length();
                }

                if (splitIndex != -1) {
                     AppLogger.getInstance().log("Parser", "⚠ Car Protocol Detected via separator [" + separator + "]");
                     
                     // 1. In this mode, the Title field actually holds the Lyric
                     finalLyric = rawTitle;

                     // 2. The Artist field holds "Song - Artist"
                     finalTitle = rawArtist.substring(0, splitIndex).trim();
                     finalArtist = rawArtist.substring(splitIndex + offset).trim();
                }
            }
        }
        
        AppLogger.getInstance().log("Repo", "✅ Posting: Title=[" + finalTitle + "] Artist=[" + finalArtist + "] Lyric=[" + (finalLyric != null ? "YES" : "NO") + "]");

        // Push Decision
        if (finalLyric != null) {
            LyricRepository.getInstance().updateLyric(finalLyric, getAppName(pkg));
        }
        LyricRepository.getInstance().updateMediaMetadata(finalTitle, finalArtist, pkg);
    }
    
    private String getAppName(String packageName) {
        if (packageName == null) return "Music";
        if (packageName.contains("qqmusic")) return "QQ Music";
        if (packageName.contains("netease")) return "NetEase";
        if (packageName.contains("miui")) return "Mi Music";
        return packageName;
    }
}

File: app/src/main/java/com/example/islandlyrics/MediaMonitorService.kt
package com.example.islandlyrics

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.media.MediaMetadata
import android.media.session.MediaController
import android.media.session.MediaSessionManager
import android.media.session.PlaybackState
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.service.notification.NotificationListenerService
import android.util.Log

class MediaMonitorService : NotificationListenerService() {

    private var mediaSessionManager: MediaSessionManager? = null
    private var componentName: ComponentName? = null
    private var prefs: SharedPreferences? = null

    private val allowedPackages = HashSet<String>()
    
    // Explicitly use fully qualified names to avoid conflicts or ambiguity if imported
    private val activeControllers = java.util.ArrayList<MediaController>()
    
    // Deduplication: Track last metadata hash to avoid processing duplicates
    private var lastMetadataHash: Int = 0

    private val sessionsChangedListener = MediaSessionManager.OnActiveSessionsChangedListener { controllers ->
        updateControllers(controllers)
    }

    private val handler = Handler(Looper.getMainLooper())
    private val stopRunnable = Runnable {
        AppLogger.getInstance().log(TAG, "Stopping service after debounce.")
        LyricRepository.getInstance().updatePlaybackStatus(false) // Logic Fix: Move status update here
        val intent = Intent(this@MediaMonitorService, LyricService::class.java)
        intent.action = "ACTION_STOP"
        startService(intent)
    }

    private val prefListener = SharedPreferences.OnSharedPreferenceChangeListener { _, key ->
        if (PREF_PARSER_RULES == key) {
            loadWhitelist()
            recheckSessions()
        } else if ("service_enabled" == key) {
            recheckSessions()
        }
    }

    override fun onCreate() {
        super.onCreate()
        prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE)
        // Load initial whitelist
        loadWhitelist()
        // Register pref listener
        prefs?.registerOnSharedPreferenceChangeListener(prefListener)
    }

    override fun onListenerConnected() {
        super.onListenerConnected()
        Log.d(TAG, "onListenerConnected")
        mediaSessionManager = getSystemService(Context.MEDIA_SESSION_SERVICE) as MediaSessionManager
        componentName = ComponentName(this, MediaMonitorService::class.java)

        mediaSessionManager?.addOnActiveSessionsChangedListener(sessionsChangedListener, componentName)

        // Initial check
        try {
            val controllers = mediaSessionManager?.getActiveSessions(componentName)
            updateControllers(controllers)
        } catch (e: SecurityException) {
             AppLogger.getInstance().log(TAG, "Security Error on Connect: ${e.message}")
        }
    }

    override fun onListenerDisconnected() {
        super.onListenerDisconnected()
        mediaSessionManager?.removeOnActiveSessionsChangedListener(sessionsChangedListener)
    }

    override fun onDestroy() {
        super.onDestroy()
        prefs?.unregisterOnSharedPreferenceChangeListener(prefListener)
    }

    private fun recheckSessions() {
        if (mediaSessionManager != null && componentName != null) {
            try {
                updateControllers(mediaSessionManager?.getActiveSessions(componentName))
            } catch (e: SecurityException) {
                AppLogger.getInstance().log(TAG, "Error refreshing sessions: ${e.message}")
            }
        }
    }

    private fun loadWhitelist() {
        // Use ParserRuleHelper to get all enabled packages (replaces old WhitelistHelper)
        val set = ParserRuleHelper.getEnabledPackages(this)
        allowedPackages.clear()
        allowedPackages.addAll(set)
        AppLogger.getInstance().log(TAG, "Whitelist updated: ${allowedPackages.size} enabled apps.")
    }

    private fun updateControllers(controllers: List<MediaController>?) {
        // CHECK MASTER SWITCH
        val isServiceEnabled = prefs?.getBoolean("service_enabled", true) ?: true
        if (!isServiceEnabled) {
            AppLogger.getInstance().log(TAG, "Master Switch OFF. Ignoring updates.")
            activeControllers.clear()
            LyricRepository.getInstance().updatePlaybackStatus(false)
            return
        }

        activeControllers.clear()
        controllers?.forEach { controller ->
            if (allowedPackages.contains(controller.packageName)) {
                activeControllers.add(controller)
                // Re-register callbacks to ensure we catch state changes
                controller.registerCallback(object : MediaController.Callback() {
                    // Debounce handler to prevent rapid/incomplete updates
                    private val debounceHandler = android.os.Handler(android.os.Looper.getMainLooper())
                    private val updateRunnable = Runnable {
                        // Double check state before updating
                        if (controller.playbackState?.state == PlaybackState.STATE_PLAYING) {
                            updateMetadataIfPrimary(controller)
                        } else {
                            // If paused/stopped, update immediately to reflect state change
                            updateMetadataIfPrimary(controller)
                        }
                    }

                    override fun onPlaybackStateChanged(state: PlaybackState?) {
                        // CRITICAL FIX: Do NOT update repository here directly!
                        // That bypasses the debounce logic in checkServiceState.
                        // Just trigger the check.
                        
                        val isPlaying = state?.state == PlaybackState.STATE_PLAYING
                        AppLogger.getInstance().log("Playback", "⏯️ State changed: ${if (isPlaying) "PLAYING" else "PAUSED/STOPPED"}")
                        
                        checkServiceState()
                    }

                    override fun onMetadataChanged(metadata: MediaMetadata?) {
                        // Add small delay (50ms) to allow notification to fully populate
                        // This helps with apps that update state before metadata (race condition)
                        debounceHandler.removeCallbacks(updateRunnable)
                        debounceHandler.postDelayed(updateRunnable, 50)
                    }
                })
            }
        }

        // Initial check
        checkServiceState()

        // Force update metadata from the CURRENT primary
        val primary = getPrimaryController()
        if (primary != null) {
            updateMetadataIfPrimary(primary)
        }
    }

    private fun checkServiceState() {
        val primary = getPrimaryController()
        val isPlaying = primary?.playbackState?.state == PlaybackState.STATE_PLAYING

        if (isPlaying) {
            // Cancel any pending stop
            handler.removeCallbacks(stopRunnable)

            // Start/Update Service
            val intent = Intent(this, LyricService::class.java)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                startForegroundService(intent)
            } else {
                startService(intent)
            }

            // Sync State
            LyricRepository.getInstance().updatePlaybackStatus(true)

        } else {
            // Debounce Stop
            handler.removeCallbacks(stopRunnable)
            handler.postDelayed(stopRunnable, 500)
            
            // LOGIC FIX: Status update moved to stopRunnable to respect debounce
        }
    }

    private fun getPrimaryController(): MediaController? {
        // Priority 1: Playing
        for (c in activeControllers) {
            val state = c.playbackState
            if (state != null && state.state == PlaybackState.STATE_PLAYING) {
                return c
            }
        }
        // Priority 2: First in list (or most recent?) - API list order usually implies recency
        if (activeControllers.isNotEmpty()) {
            return activeControllers[0]
        }
        return null // None active
    }

    private fun updateMetadataIfPrimary(controller: MediaController) {
        val primary = getPrimaryController() ?: return

        // Only update if this controller IS the primary one
        if (controller.packageName == primary.packageName) {
            val metadata = controller.metadata
            if (metadata != null) {
                extractMetadata(metadata, controller.packageName)
            }
        }
    }

    private fun extractMetadata(metadata: MediaMetadata, pkg: String) {
        val rawTitle = metadata.getString(MediaMetadata.METADATA_KEY_TITLE)
        val rawArtist = metadata.getString(MediaMetadata.METADATA_KEY_ARTIST)

        // DEDUPLICATION: Skip if metadata is identical to last update
        // Some apps (e.g., ink.trantor.coneplayer) send rapid duplicate metadata updates
        val metadataHash = java.util.Objects.hash(rawTitle, rawArtist, pkg)
        if (metadataHash == lastMetadataHash) {
            AppLogger.getInstance().log("Meta-Debug", "⏭️ Skipping duplicate metadata for [$pkg]")
            return
        }
        lastMetadataHash = metadataHash

        AppLogger.getInstance().log("Meta-Debug", "📦 RAW Data from [$pkg]:")
        AppLogger.getInstance().log("Meta-Debug", "   ↳ Title : $rawTitle")
        AppLogger.getInstance().log("Meta-Debug", "   ↳ Artist: $rawArtist")

        var finalTitle = rawTitle
        var finalArtist = rawArtist
        var finalLyric: String? = null

        // Load parser rule for this package (configurable system)
        val rule = ParserRuleHelper.getRuleForPackage(this, pkg)

        if (rule != null && rule.usesCarProtocol && rawArtist != null) {
            // Car protocol mode: Parse artist field to extract title/artist, title field contains lyric
            AppLogger.getInstance().log("Parser", "⚙️ Applying car protocol rule: separator=[${rule.separatorPattern}], order=${rule.fieldOrder}")
            
            // Parse using configurable rule
            val (parsedTitle, parsedArtist) = parseWithRule(rawArtist, rule)
            
            if (parsedTitle.isNotEmpty()) {
                finalLyric = rawTitle  // Title field holds lyric
                finalTitle = parsedTitle
                finalArtist = parsedArtist
                AppLogger.getInstance().log("Parser", "✅ Parsed: Title=[$finalTitle], Artist=[$finalArtist]")
            }
        } else if (rule != null && !rule.usesCarProtocol) {
            // Non-car-protocol mode: Use raw metadata directly (for apps like Kugou, Apple Music)
            AppLogger.getInstance().log("Parser", "ℹ️ Non-car-protocol app, using raw metadata (no lyric extraction)")
            finalTitle = rawTitle
            finalArtist = rawArtist
            // finalLyric remains null (no lyric support)
        } else {
            // No rule found, use raw data
            AppLogger.getInstance().log("Parser", "⚠️ No parser rule found for [$pkg], using raw metadata")
        }

        AppLogger.getInstance().log("Repo", "✅ Posting: Title=[$finalTitle] Artist=[$finalArtist] Lyric=[${if (finalLyric != null) "YES" else "NO"}]")

        // Push to repository
        if (finalLyric != null) {
            LyricRepository.getInstance().updateLyric(finalLyric, getAppName(pkg))
        }

        val duration = metadata.getLong(MediaMetadata.METADATA_KEY_DURATION)
        LyricRepository.getInstance().updateMediaMetadata(finalTitle ?: "Unknown", finalArtist ?: "Unknown", pkg, duration)
    }

    /**
     * Parse notification text using configurable separator and field order.
     * @param input Raw text like "Artist-Title" or "Title | Artist"
     * @param rule Parser rule with separator and field order
     * @return Pair of (title, artist)
     */
    private fun parseWithRule(input: String, rule: ParserRule): Pair<String, String> {
        val separator = rule.separatorPattern
        val splitIndex = input.indexOf(separator)
        
        if (splitIndex == -1) {
            // Separator not found, return original with empty artist
            AppLogger.getInstance().log("Parser", "⚠️ Separator [$separator] not found in: $input")
            return Pair(input, "")
        }

        // Split the input
        val part1 = input.substring(0, splitIndex).trim()
        val part2 = input.substring(splitIndex + separator.length).trim()

        // Apply field order
        return when (rule.fieldOrder) {
            FieldOrder.ARTIST_TITLE -> Pair(part2, part1)  // part1=artist, part2=title
            FieldOrder.TITLE_ARTIST -> Pair(part1, part2)  // part1=title, part2=artist
        }
    }


    private fun getAppName(packageName: String?): String {
        if (packageName == null) return "Music"
        if (packageName.contains("qqmusic")) return "QQ Music"
        if (packageName.contains("netease")) return "NetEase"
        if (packageName.contains("miui")) return "Mi Music"
        return packageName
    }

    companion object {
        private const val TAG = "MediaMonitorService"
        private const val PREFS_NAME = "IslandLyricsPrefs"
        private const val PREF_PARSER_RULES = "parser_rules_json"
    }
}
